src/onegov/search/utils.py

```diff
@@ -43,7 +42,9 @@ def searchable_sqlalchemy_models(
     )
 
 
-def filter_non_base_models(models: 'set[type[T]]') -> 'set[type[T]]':
+def filter_non_base_models(
+    models: 'set[type[T]]'
+) -> 'set[type[T]]':
```

Comment:
This is the opposite of what we want. This will now potentially miss some models rather than lead to overlapping queries. We want to get rid of models that are already covered by their base class, not the other way around.

Reply:
So instead of indexing `Topic` and `News` we only index `Page`. Same with `XYTicket` just `Ticket`

Reply:
With your examples in the tests, yes. In reality `Page` is not searchable, so it will still include both `Topic` and `News` and not `Page` (since that was never included to begin with).

I've seen some cases where polymorphic queries weren't working properly and the query always included the base class. If that's what's happening here we always need to use the polymorphic base class, regardless of whether it's searchable or not and instead apply the searchable filter on the results.

Or we manually create a filter expression on the `type` column for all the searchable polymorphic identities. There may be some performance benefit to this approach, since we avoid generating too many individual queries this way.

Reply:
With the following statement in `reindex_model` no rows will be indexed twice
```
                if i.polymorphic_on is not None:
                    q = q.filter(i.polymorphic_on == i.polymorphic_identity)
```

Reply:
I looked into it briefly:

```python
i = inspect(Topic)
i.base_mapper.class_   # Page
{
    m.polymorphic_identity
    for m in mapper.base_mapper.self_and_descendants
    if issubclass(m.class_, Searchable)
}  # {'topic', 'news'}
```

This should give you everything you need. Note that you only want to do this for models that have `i.polymorphic_on is not None`. So you want a `get_base_models` function which de-duplicates and gives you the base class for polymorphic models and just the model itself for everything else. And then you change the filter from a single polymorphic identity, to all the searchable ones.

Reply:
I see, the `base_mapper` does the trick!! I did a very similar thing which lead me to identities like `['topic', 'news', 'generic', 'extended', 'image', 'general', 'landsgemeinde', 'builtin', 'custom', None, 'FRM', 'RSV', 'EVN', 'DIR', 'CHT', 'FER', 'vacation', 'daypass', 'room', 'daily-item']`

```
+            for model in models:
+                i = inspect(model)
+                idents = set(e.polymorphic_identity
+                             for e in list(i.self_and_descendants)[1:]
+                             if e.polymorphic_identity is not None)
```